diff --git a/pjsip/src/pjsip/sip_transport.c b/pjsip/src/pjsip/sip_transport.c
index 4f483faa1..12439e3ee 100644
--- a/pjsip/src/pjsip/sip_transport.c
+++ b/pjsip/src/pjsip/sip_transport.c
@@ -2088,15 +2088,17 @@ PJ_DEF(pj_ssize_t) pjsip_tpmgr_receive_packet( pjsip_tpmgr *mgr,
              * which were sent to keep NAT bindings.
              */
             if (tmp.slen) {
-                PJ_LOG(1, (THIS_FILE, 
-                      "Error processing %d bytes packet from %s %s:%d %.*s:\n"
-                      "%.*s\n"
-                      "-- end of packet.",
+                PJ_LOG(2, (THIS_FILE,
+                      "Dropping %d bytes packet from %s %s:%d %.*s\n",
                       msg_fragment_size,
                       rdata->tp_info.transport->type_name,
-                      rdata->pkt_info.src_name, 
+                      rdata->pkt_info.src_name,
                       rdata->pkt_info.src_port,
-                      (int)tmp.slen, tmp.ptr,
+                      (int)tmp.slen, tmp.ptr));
+                PJ_LOG(4, (THIS_FILE,
+                      "Dropped packet:"
+                      "%.*s\n"
+                      "-- end of packet.",
                       (int)msg_fragment_size,
                       rdata->msg_info.msg_buf));
             }
